backend/regalloc/regalloc_split_state.ml

```diff
@@ -43,6 +43,176 @@ let log_renaming_info : indent:int -> t -> unit =
         Printmach.regset regset)
     state.phi_at_beginning
 
+(* Optimizes `destructions_at_end` and `definitions_at_beginning`, by moving
+   definitions down and destructions up. Doing so create more opportunities for
+   `RemoveReloadSpillInSameBlock`, and reduces the live ranges of temporaries,
+   hence making things slightly easier for the register allocator. *)
+module MoveSpillsAndReloads : sig
+  val optimize :
+    Cfg_with_infos.t ->
+    destructions_at_end:destructions_at_end ->
+    definitions_at_beginning:definitions_at_beginning ->
+    destructions_at_end * definitions_at_beginning
+end = struct
+  (** A definition at the begin of a block can be moved down if:
+
+     - the block does neither raise nor end with a destruction point;
+     - the block has only one (normal) successor;
+     - the successor has only one predecessor;
+     - the block and its successor are different;
+     - the defined register has no occurrences in the block. *)
+  let move_definitions_at_beginning_down :
+      Cfg_with_infos.t ->
+      definitions_at_beginning:definitions_at_beginning ->
+      definitions_at_beginning * Label.t Stack.t =
+   fun cfg_with_infos ~definitions_at_beginning ->
+    if split_debug
+    then log ~indent:1 "MoveSpillsAndReloads.move_definitions_at_beginning_down";
+    let definitions_at_beginning = ref definitions_at_beginning in
+    let doms = Cfg_with_infos.dominators cfg_with_infos in
+    let stack = Stack.create () in
+    Cfg_dominators.iter_breadth_dominator_tree
+      doms.Cfg_dominators.dominator_tree ~f:(fun (label : Label.t) ->
+        Stack.push label stack;
+        let block = Cfg_with_infos.get_block_exn cfg_with_infos label in
+        match Label.Map.find_opt label !definitions_at_beginning with
+        | None -> ()
+        | Some (definitions : Reg.Set.t) -> (
```

Comment:
They can be empty; there is a CR below
about whether such bindings could be
removed.

I don't think I understand the relation
with the fixpoint. Because we are using
the dominator tree to push down, I don't
think we can "miss" a set of definition if
if is what you mean. (And missing a set
would be suboptimal but not wrong.)